src/opentimelineio/algo/editAlgorithm.h

```diff
@@ -0,0 +1,188 @@
+//       ^
+//     | C |
+//
+//          item = item to insert (usually a clip)
+//   composition = usually a track item.
+//          time = time to insert at.  If < composition's start time, it will insert at 0 index.
+//                                     If > composition's end_time_exclusive, it will append at end.
+// remove_transitions = whether to remove transitions that intersect time.
+// fill_template = item to fill in (usually a gap),
+//                 when time > composition's time.
+void insert(
+    Item* const         item,
+    Composition*        composition,
+    RationalTime const& time,
+    bool const          remove_transitions = true,
+    Item*               fill_template      = nullptr,
+    ErrorStatus*        error_status       = nullptr);
+
+//
+// Adjust a single item's start time or duration.
+//
+// |    A    | B | C |  ->  |  A  |FILL| B | C |
+//        <--*
+//            
+//          item = Item to apply trim to (usually a clip)
+//      delta_in = RationalTime that the item's source_range().start_time()
+//                 will be adjusted by
+//     delta_out = RationalTime that the item's
+//                 source_range().end_time_exclusive() will be adjusted by
+// fill_template = item to fill in (usually a gap),
+//                 when time > composition's time.
+void trim(
+    Item*               item,
+    RationalTime const& delta_in,
+    RationalTime const& delta_out,
+    Item*               fill_template = nullptr,
+    ErrorStatus*        error_status  = nullptr);
+
+// Slice an item.
+//
+// | A | B | -> |A|A| B |
+//   ^
+// composition = usually a track item.
+//        time = time to slice at.
+void slice(
+    Composition*        composition,
+    RationalTime const& time,
+    bool const          remove_transitions = true,
+    ErrorStatus*        error_status       = nullptr);
+
+//
+// Slip an item start_time by + or -, clamping to available_range if available.
+//
+// |   A   |
+//  <----->
+//
+//          item = item to slip (usually a clip)
+//         delta = +/- rational time to slip the item by.
+void slip(Item* item, RationalTime const& delta);
+
+//
+// Slide an item start_time by + or -, moving the previous item's duration.
+// Clamps previous item's duration to available_range if available.
+//
+// | A | B | C |  ->  | A     | B | C |
+//     *--->
+//
+//          item = item to slip (usually a clip)
+//         delta = +/- rational time to slide the item by.
+void slide(Item* item, RationalTime const& delta);
+
+//
+// Adjust a source_range without affecting any other items.
+//
+// |   A   |   B   |  ->  | A |  B  |FILL|
+//      <--*
+//            
+//      item = Item to apply ripple to (usually a clip)
+//  delta_in = RationalTime that the item's source_range().start_time()
+//             will be adjusted by
+// delta_out = RationalTime that the item's
+//             source_range().end_time_exclusive() will be adjusted by
```

Comment:
what happens if the adjustment results in an element with zero or negative duration?

Reply:
It is clamped.